process/process.go

```diff
@@ -312,15 +319,19 @@ func (p *Process) handlePrevote(prevote *Prevote) {
 	}
 
 	// upon 2f+1 Prevote{currentHeight, currentRound, nil} while currentStep = StepPrevote
-	if n := p.state.Prevotes.QueryByHeightRoundBlockHash(p.state.CurrentHeight, p.state.CurrentRound, block.InvalidHash); n > 2*p.state.Prevotes.F() && p.state.CurrentStep == StepPrevote {
-		precommit := NewPrecommit(
-			p.state.CurrentHeight,
-			p.state.CurrentRound,
-			block.InvalidHash,
-		)
-		p.logger.Debugf("precommited=<nil> at height=%v and round=%v (2f+1 prevote=<nil>)", precommit.height, precommit.round)
-		p.state.CurrentStep = StepPrecommit
-		p.broadcaster.Broadcast(precommit)
+	if n := p.state.Prevotes.QueryByHeightRoundBlockHash(p.state.CurrentHeight, p.state.CurrentRound, block.InvalidHash); n > 2*p.state.Prevotes.F() {
+		if p.state.CurrentStep == StepPrevote {
+			precommit := NewPrecommit(
+				p.state.CurrentHeight,
+				p.state.CurrentRound,
+				block.InvalidHash,
+			)
+			p.logger.Debugf("precommited=<nil> at height=%v and round=%v (2f+1 prevote=<nil>)", precommit.height, precommit.round)
+			p.state.CurrentStep = StepPrecommit
+			p.broadcaster.Broadcast(precommit)
+		}
+
+		p.observer.DidReceiveSufficientNilPrevotes(p.state.Prevotes.QueryMessagesByHeightRound(p.state.CurrentHeight, p.state.CurrentRound), 2*p.state.Prevotes.F()+1)
```

Comment:
The `QueryMessagesByHeightRound` does not only return prevotes that are for `nil`. Is that going to cause a problem? Given the name of the function (`DidReceiveSufficientNilPrevotes`) it would be a little surprising that the prevotes it accepts are all necessarily for `nil`.

Reply:
I'm not sure it should be Hyperdrive's job to filter the prevotes. I think it's best to hand them all to the external application and let them filter them as they please.

Reply:
In principle, I agree. However, the method name explicitly says `Nil`. Happy to keep it this way, but should definitely be clearly documented in the interface so that there is no confusion for implementers/auditors.